src/nanoarrow/nanoarrow_types.h

```diff
@@ -32,6 +32,22 @@
 extern "C" {
 #endif
 
+#if defined _WIN32 || defined __CYGWIN__
+#if defined(NANOARROW_BUILD_DLL)
+#define NANOARROW_EXPORT __declspec(dllexport)
+#elif defined(NANOARROW_CONSUME_DLL)
+#define NANOARROW_EXPORT __declspec(dllimport)
+#else
+#define NANOARROW_EXPORT
+#endif  // defined(NANOARROW_BUILD_DLL)
+#else
+#if __GNUC__ >= 4
+#define NANOARROW_EXPORT __attribute__((visibility("default")))
+#else
+#define NANOARROW_EXPORT
+#endif  // __GNUC__ >= 4
```

Comment:
Which part of the macro are you thinking we should remove? Just the `#define NANOARROW_EXPORT` or the entire gcc block?

From what I understand, MSVC requires you to be cognizant of all the different contexts and explicitly ask for what you want. GCC manages this in a more automated fashion:

https://github.com/mesonbuild/wrapdb/pull/309#issuecomment-1042481137

Reply:
If you set gcc universally to `#define NANOARROW_EXPORT` then linkage there will fail, since we are compiling these librarys with `-fvisibility=hidden` in the Meson configuration

Reply:
I think a good way to test this would be to try the Meson subproject configuration, since that will statically link nanoarrow. Looks like that is broken on main at the moment so will get that fixed so we can more accurately compare

There definitely are symbols that are not getting exported properly (note the original commit in this PR, which was only scoped to C++); having the ability to detect these from gcc I think is helpful